lib/tapioca/dsl/compilers/rubocop.rb

```diff
@@ -0,0 +1,83 @@
+# typed: strict
+# frozen_string_literal: true
+
+begin
+  require "rubocop"
+rescue LoadError
+  return
+end
+
+module Tapioca
+  module Dsl
+    module Compilers
+      # `Tapioca::Dsl::Compilers::RuboCop` generate types for RuboCop cops.
+      # RuboCop uses macros to define methods leveraging "AST node patterns".
+      # For example, in this cop
+      #
+      #   class MyCop < Base
+      #     def_node_matcher :matches_some_pattern?, "..."
+      #
+      #     def on_send(node)
+      #       return unless matches_some_pattern?(node)
+      #       # ...
+      #     end
+      #   end
+      #
+      # the use of `def_node_matcher` will generate the method
+      # `matches_some_pattern?`, for which this compiler will generate a `sig`.
+      #
+      # More complex uses are also supported, including:
+      #
+      # - Usage of `def_node_search`
+      # - Parameter specification
+      # - Default parameter specification, including generating sigs for
+      #   `without_defaults_*` methods
+      class RuboCop < Compiler
+        ConstantType = type_member { { fixed: T.all(T.class_of(::RuboCop::Cop::Base), Extensions::RuboCop) } }
+
+        class << self
+          extend T::Sig
+          sig { override.returns(T::Enumerable[Class]) }
+          def gather_constants
+            descendants_of(::RuboCop::Cop::Base).select { |constant| name_of(constant) }
+          end
+        end
+
+        sig { override.void }
+        def decorate
+          return unless used_macros?
```

Comment:
I think you can simplify all of this with one collection:

```rb
methods = node_macros
return if methods.empty?

root.create_path(constant) do |cop_klass|
  methods.each do |name|
    create_method_from_def(cop_klass, constant.instance_method(name))
  end
end
```

Since it doesn't look like we have special behavior depending on the kind.